"Any" rewrite (read for explanation of "any")

decibelcooper opened the pull request:
# Details
It is time for a rewrite.  Let me explain...

After much input from and discussion with Alexander Kiselev about important features, I have reimagined the proio format.  Alexander is the creator of EicMC (which is usually [here](https://gitlab.com/ESC/EicMC), but it is apparently missing or made private right now, and slides can be found [here](https://www.jlab.org/indico/event/216/session/2/contribution/3/material/slides/0.pptx)), a nice Protobuf-based library for MC data exchange.  EicMC includes advanced features like multi-level catalogs for fast direct access of events.  Proio is taking more and more influence from EicMC.  In contrast, however, proio will continue to focus on having the ability to write to and read from streams (as opposed to requiring seekable output or input), and being more general than the use case of MC data exchange.

I believe that this reimagining of the proio format makes it more elegant, and removes ideas that are legacy concepts from LCIO.  The reimagined format concept is as follows (skip to Summary if you don't want the details):
* At the lowest level, data are organized into "buckets".  These are O(MB)-sized containers of some number of whole events that can either be uncompressed, or compressed with LZ4 or GZIP.  Each bucket has a header to declare the bucket size and compression type, as well as information like the number of events contained within the bucket to facilitate scanning the file.  Bucket headers can currently either be of type "EVENTS" or "FOOTER".  If the latter type, the header can be used to contain summarizing information about previous events, such as a catalog of events (there are more details here that I won't get into).  Each header is preceded by a magic number to assist recovery, and a fixed-length number declaring the size of the header message.  The buckets will be decompressed all at once to take full advantage of fast multi-threaded decompression in LZ4.
* Events are stored sequentially within the buckets, each preceded by a fixed-length number declaring the size of the protobuf message.  The protobuf message contains a list of arbitrary entries each identified by a unique ID number (which can be used as a persistent handle for the entry for referencing).  Each entry is an array of bytes, which can be any serialized protobuf message.  This is the reason for the title being `"Any" rewrite`, since this is essentially the concept of using the Any protobuf message.  A list of protobuf message types contained in the event is stored separately and referenced by each entry.  Finally, a list of "tags" is stored in the event, where each tag is identified by a human-readable string, and contains a list of entry IDs that the tag refers to.  The concept of tags replaces the concept of containers from LCIO.
* The proio streams will be fully self-describing by adding FileDescriptorProtos for all entry message types as metadata in the bucket headers as necessary.  The reader will keep an eye out for available FileDescriptorProtos in the bucket headers, and load them in the cases where they can be used for reading dynamic messages.

# Summary
Proio is morphing into a scheme that wraps Protobuf only to provide the concept of **events** which is useful for MEP and HEP.  Each event contains a set of **entries** and **tags** to describe the entries.  Entries can be any protobuf message.  Aside from that, additional features of this reimagined proio are:
* Fully self-describing
* Able to contain catalog information for the case of large files
* Organized into "buckets" for fast scanning in compressed or uncompressed cases, and efficient decompression for the case of LZ4
* Enforces persistent unique IDs for each entry in each event to facilitate referencing (e.g. calorimeter hit that references MC particles responsible for energy deposition)

Proio will otherwise strive to maintain simplicity, a limited set of features, efficient disk usage, and read performance limited primarily by protobuf parsing.

The file `proto/proio.proto` contains the protobuf messages used by the reimagined scheme and should give more of an idea of how the scheme works.  The `model` directory contains included messages that can be placed as entries, but the user is not limited to using these messages.

# Current status
* Go implementation is mostly complete.  CI is set up (#8)
* Python implementation is mostly complete.  Still needs tag manipulation tests (will create issue after merge).  CI is set up (#8)
* C++ implementation: Will create issue after merge
* Java implementation: Will create issue after merge
* Documentation: Will create issue after merge

@sbinet I know you have a lot going on, but if you have the opportunity, please give any feedback that you wish.

decibelcooper committed:
Major rewrite

Proio is being rewritten to change the way collections are defined.
Collections are now defined as a single type with generic entries that
are stored as byte arrays.  Collections have a particular type...

decibelcooper committed:
Another tweak to the way entry type names are handled

decibelcooper committed:
proto/proio.proto: fixed tabbing

decibelcooper committed:
go-proio/collection.go: Updated to take advantage of sort.Slice()

decibelcooper committed:
go-proio: removed genExtraMsgFuncs.sh

decibelcooper committed:
go-proio: rewrite further refined, and documentation added

decibelcooper committed:
go-proio/collection.go: removed redundant import declaration

decibelcooper committed:
Dump

decibelcooper committed:
Another huge overhaul.  Trying different things to find the best

decibelcooper committed:
go-proio: functional lcio2proio and tests

decibelcooper committed:
go-proio: fixes

decibelcooper committed:
go-proio: increased bucket size

decibelcooper committed:
proto: Added ability to have footer information

In order to accomodate things such as catalogs at the end of files, the
BucketHeader can now be of multiple types, one of which being "FOOTER".
The idea is that the writer can write a footer BucketHeader and then the
payload of the footer is a magic number followed by a fixed-length
number representing the size of the preceding BucketHeader protobuf
message.  A reader can then look at the end of the file and be able to
access the information contained in the footer BucketHeader.

decibelcooper committed:
.gitignore: added promc files

decibelcooper committed:
go-proio: Removed explicit model imports

decibelcooper committed:
go-proio: reader now decompresses buckets all at once

decibelcooper committed:
go-proio/lcio2proio: improved arguments for compression

decibelcooper committed:
go-proio: finished converting over tools and necessary library components

decibelcooper committed:
go-proio: added some thread safety to reader

sbinet commented:
I haven't looked at it in details yet.
but just to add entropy, let me mention these two interesting packages that I have been reminded by the move of Francesc to `source{d}`:

- https://blog.sourced.tech/post/proteus: generate `.proto` messages from Go types
- https://blog.sourced.tech/post/siva: an archive format not dissimilar to SIO and proio (except they don't really care about compression as what they archive is already somehow compressed)

decibelcooper commented:
@sbinet Thank you!

* proteus: This is a very cool concept.  However, I'm not sure that it fits with my philosophy for proio.  As proio begins to stabilize, significant emphasis will start to be placed on forwards and backwards compatibility.  To ensure this compatibility, protobuf message descriptions should be the source of truth.
* siva: Indeed, this is quite similar to this latest proio concept.  There is, however, a fundamental difference.  Siva is an archive format to the core, since it requires scanning the file backwards.  They have chosen to place block and indexing information only as footers.  Proio is intended to be able to be written and read sequentially from a stream, without having to access the EOF.

decibelcooper committed:
Travis CI: Go only, to start with

decibelcooper committed:
Ditching Make-based code generation...

in favor of code generation in language-specific build systems.

decibelcooper committed:
Code generation: moved model proto files, and implemented Go code gen

decibelcooper committed:
go-proio/reader.go: minor variable name change

decibelcooper committed:
go-proio/reader.go: fixed faulty logic in Skip()

decibelcooper committed:
py-proio: "Any" rewrite for PR #14!

Currently read only.

decibelcooper committed:
Travis CI: added py-proio...

with a single dummy test

decibelcooper committed:
.gitignore: added pytest cache directories to ignore

decibelcooper committed:
go-proio: fixed tabbing

decibelcooper committed:
go-proio: changed where compression is specified, and more testing

Compression now defaults to LZ4, and user can use SetCompression() to
change to GZIP or UNCOMPRESSED.  Testing coverage for the three
compression options was also improved.

decibelcooper committed:
py-proio/setup.py: now discovers model modules, rather than hardcoding

decibelcooper committed:
go-proio: better Event.tagEntry() and proio-summary

decibelcooper committed:
py-proio: functional write + read

decibelcooper committed:
py-proio: added __str__() function to Event

decibelcooper committed:
py-proio: bug fixes and basic tests

decibelcooper committed:
.travis.yml: updated python deps during install

decibelcooper committed:
.travis.yml: updated pytest duing install

decibelcooper committed:
py-proio: rearranged model modules to not make protobuf module angry

protobuf may check to make sure that modules creating descriptors
directly are generated, and it does this by checking that the file name
ends with '_pb2.py'.  I don't want the import path to have this '_pb2',
so I renamed the files, which caused an exception.  Now, the modules are
wrapped in an '__init__.py'.

decibelcooper committed:
py-proio: fixed sort() call in Event

decibelcooper committed:
py-proio: added dummy model module for python2

decibelcooper committed:
go-proio: added more tests and added more tag methods to Event

decibelcooper committed:
go-proio: fixed broken tools after last commit

decibelcooper committed:
.travis.yml: added branch safelist

decibelcooper committed:
go-proio tests: fixed Error calls that needed to be Errorf

decibelcooper committed:
py-proio/README.md: added --upgrade to pip install example

decibelcooper committed:
py-proio/setup.py: fixed after f9d32cde329725243c43ebca493652651b8a53aa

decibelcooper committed:
py-proio/writer.py: fixed multi-bucket writes

decibelcooper committed:
py-proio: fixed Event.add_entries()

decibelcooper committed:
py-proio/tests: added another ref_deref test

decibelcooper committed:
README.md: added Travis CI build status image

decibelcooper committed:
Cleared cpp and java directories for now
